---
title: "Phylogenetic Analysis of Salamander Genome Size"
author: "Clay Cressler"
date: "06/09/2020"
output: bookdown::html_document2
self_contained: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev=c('png','tiff'),
                      fig.path='figures/')
library(ape)
library(ouch)
library(OUwie)
library(tidyverse)
library(parallel)

```

This is an updated version of the analyses that were initially carried out several years ago. 
We had wanted to use $\texttt{Ouwie}$ to do these analyses but previously, there were undiagnosed issues with its fitting algorithm.
As such, we ended up using $\texttt{ouch}$ for all analyses, which prevented us from ever testing models with multiple $\alpha$ or $\sigma^2$ parameters.
However, I spent time working through these issues in Spring 2020 and am now satisfied that $\texttt{ouch}$ and $\texttt{OUwie}$ produce identical results, at least once you specify a value for 'shift.point' to make the way that $\texttt{OUwie}$ handles regimes identical to $\texttt{ouch}$.

```{r, echo=FALSE}
tree <- read.nexus("av_ultra_fulldataset.nex")
tree_metamorphosis <- tree_mpd <- tree_mpdMpleth <- tree
## Load the hypotheses
nodedata <- read.csv('ouwie_nodelabels.csv')
nodedata <- nodedata[,c('node','metamorphosis','mpd','mpdMpleth')]
## metamorphosis has two regimes: M, x
## mpd has three regimes: D, M, P
## mpdMpleth has four regimes: D, M, Mpleth, P
## code Regime names as numbers for the OUwie algorithm
nn <- nodedata
nn[-1] <- sapply(nn[-1], as.character)
nn[nn=="D"] <- 1
nn[nn=="M"] <- 2
nn[nn=="Mpleth"] <- 3
nn[nn=="P"] <- 4
nn[nn=="x"] <- 5
## Add node labels to trees
tree_metamorphosis$node.label <- nn$metamorphosis
tree_mpd$node.label <- nn$mpd
tree_mpdMpleth$node.label <- nn$mpdMpleth
## Tip data
dat <- read.csv("tree1.dat.csv")
dat$genomesize <- log(dat$genomesize)
tdat <- dat[!is.na(dat$genomesize),]  # tip data
ndat <- dat[is.na(dat$genomesize),]   # node data
## reorder tdat to match order of tip labels in phylo tree
oo <- sapply(tree$tip.label, function(x) grep(x, tdat$labels))
tipdat <- tdat[oo,]
tipdat <- tipdat[c("labels", "metamorphosis", "mpd", "mpdMpleth", "genomesize")]
## code Regime names as numbers for the OUwie algorithm
tt <- tipdat
tt[,1:(ncol(tipdat)-1)] <- sapply(tt[,1:(ncol(tipdat)-1)], as.character)
tt[tt=="D"] <- 1
tt[tt=="M"] <- 2
tt[tt=="Mpleth"] <- 3
tt[tt=="P"] <- 4
tt[tt=="x"] <- 5
### input dataframes - one for each hypothesis
data.metamorphosis <- tt[c("labels", "metamorphosis", "genomesize")]
data.mpd <- tt[c("labels", "mpd", "genomesize")]
data.mpdMpleth <- tt[c("labels", "mpdMpleth", "genomesize")]


##############################################################
## OUCH setup
##############################################################
dat1 <- read.csv("tree1.dat.csv")
tree1 <- read.nexus("av_ultra_fulldataset.nex")
tree <- ape2ouch(tree1)
dat <- dat1
rownames(dat) <- dat$nodes
regimes <- dat[c("metamorphosis", "mpd", "mpdMpleth","OU1","fivereg")]
gsz <- log(dat['genomesize'])

```

```{r, echo=FALSE, eval=FALSE}
## Single-rate BM (this is hypothesis-independent)
bm.ouwie <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("BM1"),root.station=TRUE,scaleHeight=TRUE,quiet=TRUE)
bm.ouch <- brown(gsz, tree)

summary(bm.ouch)$sigma.squared %>% as.numeric
bm.ouwie$solution["sigma.sq",]

summary(bm.ouch)$theta %>% as.numeric
bm.ouwie$theta[1]

```

Here are the three different adaptive hypothesis regime paintings (Fig. \@ref(fig:hyp)). 

```{r, echo=FALSE, eval=FALSE}
## Fit the mpdMpleth hypothesis
mpdMpleth.ouwie <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.ouch <- hansen(gsz, tree, regimes['mpdMpleth'], sqrt.alpha=1, sigma=1)

## Compare sigma.sq estimates
summary(mpdMpleth.ouch)$sigma.squared %>% as.numeric
mpdMpleth.ouwie$solution["sigma.sq",1]

## Compare alpha estimates
summary(mpdMpleth.ouch)$alpha %>% as.numeric
mpdMpleth.ouwie$solution["alpha",1]

## Compare optima estimates
summary(mpdMpleth.ouch)$optima$genomesize
mpdMpleth.ouwie$theta[,1]

## Compare the likelihoods
summary(hansen(gsz, tree=tree, regimes=regimes['mpdMpleth'], sqrt.alpha=sqrt(mpdMpleth.ouwie$solution["alpha",1]), sigma=sqrt(mpdMpleth.ouwie$solution["sigma.sq",1]), fit=FALSE))$loglik
mpdMpleth.ouwie$loglik


```

```{r hyp, echo=FALSE, fig.height=10, fig.width=10, fig.cap="The three evolutionary hypotheses, specified as regime paintings. Black = Direct development, Red = Metamorphosis, Green = Plethodontid metamorphosis, Blue = Paedomorphosis, Cyan = Direct development or paedomorphosis."}

#png(file="~/Desktop/hypotheses.png", width=5, height=7, units='in', res=300)
par(mar=c(0.5,0.5,2,0.5), oma=c(0.5,0.5,0.5,0.5))
layout(matrix(c(rep(1,1),rep(2,1), rep(3,3)), 1, 5, byrow=TRUE))
plot(tree_metamorphosis, show.tip.label=FALSE)
mtext("metamorphosis", side=3, line=0, cex=0.75)
nodelabels(pch=21, bg=tree_metamorphosis$node.label)

plot(tree_mpd, show.tip.label=FALSE)
mtext("meta-paed-dd", side=3, line=0, cex=0.75)
nodelabels(pch=21, bg=tree_mpd$node.label)

plot(tree_mpdMpleth)
mtext("meta2-paed-dd            ", side=3, line=0, cex=0.75)
nodelabels(pch=21, bg=tree_mpdMpleth$node.label)
#dev.off()

```


```{r, echo=FALSE, eval=FALSE}
## Fit the mpd hypothesis
mpd.ouwie <- OUwie(tree_mpd,data.mpd,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.ouch <- hansen(gsz, tree, regimes['mpd'], sqrt.alpha=1, sigma=1)

## Compare sigma.sq estimates
summary(mpd.ouch)$sigma.squared %>% as.numeric
mpd.ouwie$solution["sigma.sq",1]

## Compare alpha estimates
summary(mpd.ouch)$alpha %>% as.numeric
mpd.ouwie$solution["alpha",1]

## Compare optima estimates
summary(mpd.ouch)$optima$genomesize
mpd.ouwie$theta[,1]

## Compare the likelihoods
summary(hansen(gsz, tree=tree, regimes=regimes['mpd'], sqrt.alpha=sqrt(mpd.ouwie$solution["alpha",1]), sigma=sqrt(mpd.ouwie$solution["sigma.sq",1]), fit=FALSE))$loglik
mpd.ouwie$loglik
```

```{r, echo=FALSE, eval=FALSE}
## Fit the metamorphosis hypothesis
metamorphosis.ouwie <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.ouch <- hansen(gsz, tree, regimes['metamorphosis'], sqrt.alpha=1, sigma=1)

## Compare sigma.sq estimates
summary(metamorphosis.ouch)$sigma.squared %>% as.numeric
metamorphosis.ouwie$solution["sigma.sq",1]

## Compare alpha estimates
summary(metamorphosis.ouch)$alpha %>% as.numeric
metamorphosis.ouwie$solution["alpha",1]

## Compare optima estimates
summary(metamorphosis.ouch)$optima$genomesize
metamorphosis.ouwie$theta[,1]

## Compare the likelihoods
summary(hansen(gsz, tree=tree, regimes=regimes['metamorphosis'], sqrt.alpha=sqrt(metamorphosis.ouwie$solution["alpha",1]), sigma=sqrt(metamorphosis.ouwie$solution["sigma.sq",1]), fit=FALSE))$loglik
metamorphosis.ouwie$loglik
```

Since I have confirmed that it is possible to get nearly identical results between $\texttt{OUwie}$ and $\texttt{ouch}$, I will proceed with fitting all of the other multi-rate models: a Brownian motion model where the drift parameter ($\sigma$) varies across regimes (BMS), an Ornstein-Uhlenbeck (OU) model where selection strength varies across regimes ($\alpha$) (OUMA), an OU model where drift varies across regimes (OUMV), and an OU model where both selection and drift vary across regimes (OUMVA). 

```{r, echo=FALSE, eval=FALSE}
## Fit all of the model variants (BMS, OUMA, OUMV, OUMVA) to each hypothesis
metamorphosis.OUM <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.BMS <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMA <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMV <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMVA <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

mpd.OUM <- OUwie(tree_mpd,data.mpd,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.BMS <- OUwie(tree_mpd,data.mpd,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMA <- OUwie(tree_mpd,data.mpd,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMV <- OUwie(tree_mpd,data.mpd,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMVA <- OUwie(tree_mpd,data.mpd,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

mpdMpleth.OUM <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.BMS <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMA <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMV <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMVA <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

## save these results
results <- list(BM1=bm.ouwie,
                metamorphosis.OUM=metamorphosis.OUM,
                metamorphosis.BMS=metamorphosis.BMS,
                metamorphosis.OUMA=metamorphosis.OUMA,
                metamorphosis.OUMV=metamorphosis.OUMV,
                metamorphosis.OUMVA=metamorphosis.OUMVA,
                mpd.OUM=mpd.OUM,
                mpd.BMS=mpd.BMS,
                mpd.OUMA=mpd.OUMA,
                mpd.OUMV=mpd.OUMV,
                mpd.OUMVA=mpd.OUMVA,
                mpdMpleth.OUM=mpdMpleth.OUM,
                mpdMpleth.BMS=mpdMpleth.BMS,
                mpdMpleth.OUMA=mpdMpleth.OUMA,
                mpdMpleth.OUMV=mpdMpleth.OUMV,
                mpdMpleth.OUMVA=mpdMpleth.OUMVA
                )
saveRDS(results, "all_model_fits.RDS")

```

```{r, echo=FALSE}
results <- readRDS("all_model_fits.RDS")
```

The best-fitting model is the OU model with 4 selective regimes (1=direct development, 2=metamorphosis, 3=metamorphosing plethodontids, 4=paedomorphosis), with separate drift parameters for each regime, although the model with a single drift parameter has very similar AICc score. Here are the AICc values and parameter estimates for each of the models.

```{r, echo=FALSE}
data.frame(Hypothesis = lapply(results, function(x) x$AICc) %>% unlist %>% names,
           loglik = lapply(results, function(x) x$loglik) %>% unlist %>% signif(3),
           AICc = lapply(results, function(x) x$AICc) %>% unlist %>% signif(3),
           alpha = lapply(results, function(x) unique(x$solution['alpha',]) %>% signif(3)) %>% unlist,
           sigma = lapply(results, function(x) unique(x$solution['sigma.sq',]) %>% sqrt %>% signif(3) %>% paste(., collapse=", ")) %>% unlist,
           theta = lapply(results, function(x) x$theta[,1] %>% signif(3) %>% paste(., collapse=", ")) %>% unlist
           ) %>% 
  arrange(AICc)
```

To help interpret the biological implications of the parameter values of the best-fitting model, it is useful to calculate the values of the selection strength and drift terms of the OU model ($\alpha (\theta-X)$ and $\sigma$).
I will assume that $X$ is the average genome size for salamanders in each of the selective regimes. 
Plugging those genome size values, along with the corresponding parameter estimates for each regime gives a sense of how strong selection and drift are in each regime.
Based on this data, you can say genome size for the direct developers and the metamorphosers is primarily driven by drift, whereas metamorphosing plethodontids appear to be under fairly strong selection for small genome size and paedomorphs for large genome size. 
Or, to put it another way, trait change in direct developers and metamorphosers is dominated by random processes, but there is a strong deterministic trend towards small genomes in metamorphosing plethodontids, and towards large genomes in paedomorphs.

```{r, echo=FALSE}
data.mpdMpleth %>% group_by(mpdMpleth) %>% summarise(meanGenomeSize=signif(mean(genomesize),3)) %>% 
  mutate(., 
         alpha=results[["mpdMpleth.OUMV"]]$solution['alpha',] %>% signif(3),
         theta=results[["mpdMpleth.OUMV"]]$theta[,1] %>% signif(3),
         'selection strength'=(alpha*(theta-meanGenomeSize)) %>% signif(3),
         'drift intensity'=sqrt(results[["mpdMpleth.OUMV"]]$solution['sigma.sq',]) %>% signif(3))
                    
```
However, it is important to assess how strong the support for this hypothesis truly is. 
Following the protocol of Boettiger et al. (2012) for phylogenetic Monte Carlo, I need to simulate datasets under different models and ask how often the true (data-generating) model is rejected in favor of an alternative model.
We will do the following pairwise comparisons of models: 

1. mpdMpleth (OUMV) versus BM1: this tests whether we can confidently reject a non-adaptive hypothesis;

2. mpdMpleth (OUMV) versus mpdMpleth (OUM): this tests whether multiple $\sigma$ values is actually well-supported;

3. mpdMpleth (OUMV) versus mpd (OUM): this tests whether a separate regime for metamorphosing plethodontids is supported;

4. mpdMpleth (OUMV) versus metamorphosis (OUM): this tests whether distinguishing between differnet non-metamorphosis strategies is supported.

We will simulate datasets under each of the 5 models above, and then fit data generated by each of the two models under comparison to each of those models.


```{r, eval=FALSE, echo=FALSE}

## Generate 1000 simulated datasets under each of the 5 models of interest
PMCdata <- list(length=1000)
for (i in 1:1000) {
  print(i)
  ## simulate a dataset with each of the 5 models under consideration
  ## Brownian motion
  OUwie.sim(tree_mpdMpleth, 
            data.mpdMpleth, 
            shift.point=1, 
            alpha=rep(1e-10,4),
            sigma.sq=rep(results[["BM1"]]$solution['sigma.sq',] %>% as.numeric,4),
            theta0=results[["BM1"]]$theta[1,1] %>% as.numeric,
            theta=rep(results[["BM1"]]$theta[1,1] %>% as.numeric,4),
            ) -> bm1.data
  ## OUM under the mpdMpleth hypothesis
  OUwie.sim(tree_mpdMpleth, 
            data.mpdMpleth, 
            shift.point=1, 
            alpha=results[["mpdMpleth.OUM"]]$solution['alpha',] %>% as.numeric,
            sigma.sq=results[["mpdMpleth.OUM"]]$solution['sigma.sq',] %>% as.numeric,
            theta0=results[["mpdMpleth.OUM"]]$theta[2,1] %>% as.numeric,
            theta=results[["mpdMpleth.OUM"]]$theta[,1] %>% as.numeric) -> oum.mpdMpleth.data
  ## Note: OUwie.sim rewrites the names of the regimes, so fix that problem now or when you do any 
  ## refitting there will be problems (in particular, tree_mpdMpleth and oum.mpdMpleth.data have different
  ## names for the regimes represented on the internal nodes and the tips, respectively)
  oum.mpdMpleth.data[,2] <- data.mpdMpleth[,2]
  ## OUMV under the mpdMpleth hypothesis
  OUwie.sim(tree_mpdMpleth, 
            data.mpdMpleth, 
            shift.point=1, 
            alpha=results[["mpdMpleth.OUMV"]]$solution['alpha',] %>% as.numeric,
            sigma.sq=results[["mpdMpleth.OUMV"]]$solution['sigma.sq',] %>% as.numeric,
            theta0=results[["mpdMpleth.OUMV"]]$theta[2,1] %>% as.numeric,
            theta=results[["mpdMpleth.OUMV"]]$theta[,1] %>% as.numeric) -> oumv.mpdMpleth.data
  oumv.mpdMpleth.data[,2] <- data.mpdMpleth[,2]
  ## OUM under the mpd hypothesis
  OUwie.sim(tree_mpd, 
            data.mpd, 
            shift.point=1, 
            alpha=results[["mpd.OUM"]]$solution['alpha',] %>% as.numeric,
            sigma.sq=results[["mpd.OUM"]]$solution['sigma.sq',] %>% as.numeric,
            theta0=results[["mpd.OUM"]]$theta[2,1] %>% as.numeric,
            theta=results[["mpd.OUM"]]$theta[,1] %>% as.numeric) -> oum.mpd.data
  oum.mpd.data[,2] <- data.mpd[,2]
  ## OUM under the meta hypothesis
  OUwie.sim(tree_metamorphosis, 
            data.metamorphosis, 
            shift.point=1, 
            alpha=results[["metamorphosis.OUM"]]$solution['alpha',] %>% as.numeric,
            sigma.sq=results[["metamorphosis.OUM"]]$solution['sigma.sq',] %>% as.numeric,
            theta0=results[["metamorphosis.OUM"]]$theta[1,1] %>% as.numeric,
            theta=results[["metamorphosis.OUM"]]$theta[,1] %>% as.numeric) -> oum.meta.data
  oum.meta.data[,2] <- data.metamorphosis[,2]
  PMCdata[[i]] <- list(bm1.data=bm1.data,
                       oum.mpdMpleth.data=oum.mpdMpleth.data,
                       oumv.mpdMpleth.data=oumv.mpdMpleth.data,
                       oum.mpd.data=oum.mpd.data,
                       oum.meta.data=oum.meta.data)
}
saveRDS(PMCdata, file="PMCdata.RDS")

## Now, with all of this data fit the competing models according to the following pairwise comparisons of interest:
## 1. mpdMpleth (OUMV) versus BM1: this tests whether we can confidently reject a non-adaptive hypothesis;
## 2. mpdMpleth (OUMV) versus mpdMpleth (OUM): this tests whether multiple $\sigma$ values is actually well-supported;
## 3. mpdMpleth (OUMV) versus mpd (OUM): this tests whether a separate regime for metamorphosing plethodontids is supported;
## 4. mpdMpleth (OUMV) versus metamorphosis (OUM): this tests whether distinguishing between differnet non-metamorphosis strategies is supported.

fitComparison <- function(data) {
  
  #########################################################
  ## Pairwise comparision 1. mpdMpleth (OUMV) versus BM1 ##
  #########################################################
  
  ## Fit the oumv.mpdMpleth data to the mpdMpleth hypothesis under an OUMV model (the true data-generating model)
  fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data <- OUwie(tree_mpdMpleth,data[["oumv.mpdMpleth.data"]],model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the oumv.mpdMpleth data to the BM1 hypothesis
  fit.bm1.to.oumv.mpdMpleth.data <- OUwie(tree_mpdMpleth,data[["oumv.mpdMpleth.data"]],model=c("BM1"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the bm1 data to the mpdMpleth hypothesis under an OUMV model
  bm1.as.mpdMpleth <- data[["bm1.data"]]
  bm1.as.mpdMpleth[,2] <- data[["oumv.mpdMpleth.data"]][,2]
  fit.oumv.mpdMpleth.to.bm1.data<- OUwie(tree_mpdMpleth,bm1.as.mpdMpleth,model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the bm1 data to the bm1 hypothesis (the true data-generating model)
  fit.bm1.to.bm1.data <- OUwie(tree_mpdMpleth, data[["bm1.data"]], model=c("BM1"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ##################################################################
  ## Pairwise comparision 2. mpd (OUM) versus metamorphosis ##
  ##################################################################
  
  ## Fit the metamorphosis (OUM) model to the oum.mpd data
  mpd.as.meta <- data[["oum.mpd.data"]]
  mpd.as.meta[,2] <- data[["oum.meta.data"]][,2]
  fit.oum.meta.to.oum.mpd.data <- OUwie(tree_metamorphosis, mpd.as.meta, model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpd (OUM) model to the oum.meta data
  meta.as.mpd <- data[["oum.meta.data"]]
  meta.as.mpd[,2] <- data[["oum.mpd.data"]][,2]
  fit.oum.mpd.to.oum.meta.data <- OUwie(tree_mpd, meta.as.mpd, model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the metamorphosis (OUM) model to the oum.meta data (the true data-generating model)
  fit.oum.meta.to.oum.meta.data <- OUwie(tree_metamorphosis, data[["oum.meta.data"]], model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)

  ##################################################################
  ## Pairwise comparision 3. mpdMpleth (OUMV) versus metamorphosis ##
  ##################################################################
  
  ## Fit the metamorphosis (OUM) model to the oumv.mpdMpleth data
  mpdMpleth.as.meta <- data[["oumv.mpdMpleth.data"]]
  mpdMpleth.as.meta[,2] <- data[["oum.meta.data"]][,2]
  fit.oum.meta.to.oumv.mpdMpleth.data <- OUwie(tree_metamorphosis, mpdMpleth.as.meta, model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpdMpleth (OUMV) model to the oum.meta data
  meta.as.mpdMpleth <- data[["oum.meta.data"]]
  meta.as.mpdMpleth[,2] <- data[["oumv.mpdMpleth.data"]][,2]
  fit.oumv.mpdMpleth.to.oum.meta.data <- OUwie(tree_mpdMpleth, meta.as.mpdMpleth, model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ###############################################################
  ## Pairwise comparision 4. mpdMpleth (OUMV) versus mpd (OUM) ##
  ###############################################################
 
  ## Fit the mpd (OUM) model to the oumv.mpdMpleth data
  mpdMpleth.as.mpd <- data[["oumv.mpdMpleth.data"]]
  mpdMpleth.as.mpd[,2] <- data[["oum.mpd.data"]][,2]
  fit.oum.mpd.to.oumv.mpdMpleth.data <- OUwie(tree_mpd, mpdMpleth.as.mpd, model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpdMpleth (OUMV) model to the oum.mpd data
  mpd.as.mpdMpleth <- data[["oum.mpd.data"]]
  mpd.as.mpdMpleth[,2] <- data[["oumv.mpdMpleth.data"]][,2]
  fit.oumv.mpdMpleth.to.oum.mpd.data <- OUwie(tree_mpdMpleth, mpd.as.mpdMpleth, model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpd (OUM) model to the oum.mpd data (the true data-generating model)
  fit.oum.mpd.to.oum.mpd.data <- OUwie(tree_mpd, data[["oum.mpd.data"]], model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  #####################################################################
  ## Pairwise comparision 5. mpdMpleth (OUMV) versus mpdMpleth (OUM) ##
  #####################################################################
  
  ## Fit the mpdMpleth hypothesis (OUM) model to the oumv.mpdMpleth data
  fit.oum.mpdMpleth.to.oumv.mpdMpleth.data <- OUwie(tree_mpdMpleth,data[["oumv.mpdMpleth.data"]],model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpdMpleth hypothesis (OUMV) model to the oum.mpdMpleth data
  fit.oumv.mpdMpleth.to.oum.mpdMpleth.data <- OUwie(tree_mpdMpleth,data[["oum.mpdMpleth.data"]],model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpdMpleth hypothesis (OUM) model to the oum.mpdMpleth data (the true data-generating model)
  fit.oum.mpdMpleth.to.oum.mpdMpleth.data <- OUwie(tree_mpdMpleth,data[["oum.mpdMpleth.data"]],model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  #########################################################################

  list(## Comparison #1
       fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data=fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data,
       fit.bm1.to.oumv.mpdMpleth.data=fit.bm1.to.oumv.mpdMpleth.data,
       fit.oumv.mpdMpleth.to.bm1.data=fit.oumv.mpdMpleth.to.bm1.data,
       fit.bm1.to.bm1.data=fit.bm1.to.bm1.data,
       ## Comparison #2
       fit.oum.mpd.to.oum.meta.data=fit.oum.mpd.to.oum.meta.data,
       fit.oum.meta.to.oum.mpd.data=fit.oum.meta.to.oum.mpd.data,
       fit.oum.meta.to.oum.meta.data=fit.oum.meta.to.oum.meta.data,
       ## Comparison #3
       fit.oum.meta.to.oumv.mpdMpleth.data=fit.oum.meta.to.oumv.mpdMpleth.data,
       fit.oumv.mpdMpleth.to.oum.meta.data=fit.oumv.mpdMpleth.to.oum.meta.data,
       ## Comparison #4
       fit.oum.mpd.to.oumv.mpdMpleth.data=fit.oum.mpd.to.oumv.mpdMpleth.data,
       fit.oumv.mpdMpleth.to.oum.mpd.data=fit.oumv.mpdMpleth.to.oum.mpd.data,
       fit.oum.mpd.to.oum.mpd.data=fit.oum.mpd.to.oum.mpd.data,
       ## Comparison #5
       fit.oum.mpdMpleth.to.oumv.mpdMpleth.data=fit.oum.mpdMpleth.to.oumv.mpdMpleth.data,
       fit.oumv.mpdMpleth.to.oum.mpdMpleth.data=fit.oumv.mpdMpleth.to.oum.mpdMpleth.data,
       fit.oum.mpdMpleth.to.oum.mpdMpleth.data=fit.oum.mpdMpleth.to.oum.mpdMpleth.data,
       )
}

mclapply(PMCdata[1:500],
         fitComparison,
         mc.cores=4) -> PMCfits
saveRDS(PMCfits, file="PMCfits.RDS")




```

We can also use these fits to look at the confidence intervals for each of the parameters of the best-fitting model using a parametric bootstrap.

```{r}
PMCfits <- readRDS("PMCfits.RDS")
  
## Bootstrap parameter estimates
data.frame(parameter=c('alpha','sigma.sq (DD)', 'sigma.sq (M)', 'sigma.sq (Mpleth)', 'sigma.sq (P)', 'theta (DD)', 'theta (M)', 'theta (Mpleth)', 'theta (P)'),
           CI=c((lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$solution['alpha',] %>% unique) %>% unlist %>% sort)[c(12,488)] %>% signif(3) %>% paste(., collapse=", "),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$solution['sigma.sq',1]) %>% unlist %>% sort)[c(12,488)] %>% sqrt %>% signif(3) %>% paste(., collapse=", "),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$solution['sigma.sq',2]) %>% unlist %>% sort)[c(12,488)] %>% sqrt %>% signif(3) %>% paste(., collapse=", "),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$solution['sigma.sq',3]) %>% unlist %>% sort)[c(12,488)] %>% sqrt %>% signif(3) %>% paste(., collapse=", "),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$solution['sigma.sq',4]) %>% unlist %>% sort)[c(12,488)] %>% sqrt %>% signif(3) %>% paste(., collapse=", "),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$theta[1,1]) %>% unlist %>% sort)[c(12,488)] %>% signif(3) %>% paste(., collapse=","),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$theta[2,1]) %>% unlist %>% sort)[c(12,488)] %>% signif(3) %>% paste(., collapse=","),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$theta[3,1]) %>% unlist %>% sort)[c(12,488)] %>% signif(3) %>% paste(., collapse=","),
                (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$theta[4,1]) %>% unlist %>% sort)[c(12,488)] %>% signif(3) %>% paste(., collapse=","))
)
```


To determine whether the data really supports the conclusions we want to draw from the AICc table, we use the phylogenetic Monte Carlo approach of Boettiger et al. 2012. 
We define the difference in the log-likelihoods between the two models as our test statistic: $\delta  = -2 (\log L_0 - \log L_1)$, where we assume that the $L_1$ is the likelihood of the more complex model.
Higher values of this likelihood ratio indicate more support for the complex model.
We will use the simulated datasets to estimate the distribution of $\delta$ when the data is generated by either of the two models under comparison.
Our first comparison is between the mpdMpleth adaptive hypothesis, fitted with the OUMV model, and the non-adaptive Brownian motion hypothesis.
The p-value for this comparison is essentially 0: when the data is generated by a non-adaptive model, the value of the test statistic is *never* as large as the value that we observed with the true dataset.

```{r}
## The true value of the test statistic
delta <- -2*(results[["BM1"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)

## The distribution of the test statistic when the data is generated by the BM1 model
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.bm1.to.bm1.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.bm1.data"]]$loglik) %>% unlist))

## the proportion of the simulated values larger than the test statistic provides an approximation to the p-value for the test, the probability that a difference at least as large would be seen under model 0 (BM)
sum(delta < delta.dist)/length(delta.dist)

```
We can also look at the overlap between the distributions of the test statistic when the data is generated by the non-adaptive model versus the adaptive model to get a sense of power. You can see that the power is very high (0.994), meaning that 99.4\% of the test statistic values when the data is generated by the adaptive model are larger than the 95th percentile of the distribution when the data is generated by the non-adaptive model. Both the p-value and the power can be visualized by a plot of the two distributions  (Fig. \@ref(fig:pmcResults)A).)

```{r}
## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.bm1.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
## The degree of overlap between this distribution and the one above gives a sense of the power.
## The amount of this distribution to the right of the test statistic value that would reject the non-adaptive model with a false positive rate of 5% above gives the probability of rejecting model 0 when the data are produced by model 1 (e.g., the power)
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```
```{r pmcResults, echo=FALSE, fig.height=7, fig.width=6, fig.cap="Comparing the distributions of the test statistic when the data is generated by simpler (light gray) or more complex (dark gray) evolutionary models. The dashed line gives the actual value of the test statistic for the true data."}
par(mar=c(5,5,1,1), oma=rep(0,4), mfrow=c(3,2))
## BM1 versus mpdMpleth OUMV
delta <- -2*(results[["BM1"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.bm1.to.bm1.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.bm1.data"]]$loglik) %>% unlist))
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.bm1.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('brownian','meta2-dd-paed'), fill=c(gray(0.7),gray(0.2)), bty='n', cex=0.75)
legend(x=min(c(d1$x,d2$x)), y=max(c(d1$y,d2$y)), "A", bty='n', xjust=0.5, yjust=0.5)

## meta versus mpd
delta <- -2*(results[["metamorphosis.OUM"]]$loglik - results[["mpd.OUM"]]$loglik)
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oum.meta.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oum.meta.data"]]$loglik) %>% unlist))
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oum.mpd.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oum.mpd.data"]]$loglik) %>% unlist))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('metamorphosis','meta-dd-paed'), fill=c(gray(0.7),gray(0.2)), bty='n', cex=0.75)
legend(x=min(c(d1$x,d2$x)), y=max(c(d1$y,d2$y)), "B", bty='n', xjust=0.5, yjust=0.5)

## meta versus mpdMpleth (OUMV)
delta <- -2*(results[["metamorphosis.OUM"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oum.meta.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oum.meta.data"]]$loglik) %>% unlist))
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('metamorphosis','meta2-dd-paed'), fill=c(gray(0.7),gray(0.2)), bty='n', cex=0.75)
legend(x=min(c(d1$x,d2$x)), y=max(c(d1$y,d2$y)), "C", bty='n', xjust=0.5, yjust=0.5)


## mpd versus mpdMpleth OUMV
delta <- -2*(results[["mpd.OUM"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oum.mpd.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oum.mpd.data"]]$loglik) %>% unlist))
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('meta-dd-paed','meta2-dd-paed'), fill=c(gray(0.7),gray(0.2)), bty='n', cex=0.75)
legend(x=min(c(d1$x,d2$x)), y=max(c(d1$y,d2$y)), "D", bty='n', xjust=0.5, yjust=0.5)


##  mpdMpleth OUM versus mpdMpleth OUMV
delta <- -2*(results[["mpdMpleth.OUM"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpdMpleth.to.oum.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oum.mpdMpleth.data"]]$loglik) %>% unlist))
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('meta2-dd-paed (OUM)','meta2-dd-paed (OUMV)'), fill=c(gray(0.7),gray(0.2)), bty='n', cex=0.75)
legend(x=min(c(d1$x,d2$x)), y=max(c(d1$y,d2$y)), "E", bty='n', xjust=0.5, yjust=0.5)

```

We can repeat these analyses to look at the p-value and the power for the test for whether there is support for separate constraints for paedomorphs versus direct developers (comparing metamorphosis and mpd). The p-value (0.122) and power (0.434) suggest that there is only weak evidence for this hypothesis (Fig. \@ref(fig:pmcResults)B).

```{r}
## The true value of the test statistic
delta <- -2*(results[["metamorphosis.OUM"]]$loglik - results[["mpd.OUM"]]$loglik)
## The distribution of the test statistic when the data is generated by the oum.meta model
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oum.meta.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oum.meta.data"]]$loglik) %>% unlist))
## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oum.mpd model
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oum.mpd.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oum.mpd.data"]]$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```

We can examine the p-value and the power for the third comparison, which determines how much support there is distinguishing among different metamorphosis strategies (comparing oum.meta to oumv.mpdMpleth). Here we see a fairly low p-value (0.038) and a fairly high power (0.87), and less overlap between the two distributions  (Fig. \@ref(fig:pmcResults)C). This suggests that the improved explanatory power of the best model is driven primarily by the inclusion of non-feeding metamorphosers as a separate regime. We could test that hypothesis more explicitly by creating a new regime painting that includes non-metamorphosers as a single regime and splits out feeding and non-feeding metamorphosers into two regimes.

```{r}
## The true value of the test statistic
delta <- -2*(results[["metamorphosis.OUM"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)

## The distribution of the test statistic when the data is generated by the oum.mpd model
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oum.meta.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oum.meta.data"]]$loglik) %>% unlist))

## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.meta.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```

We can also examine the p-value and the power for the fourth comparison, which determines how much support there is for metamorphosing plethodontids being a unique selective regime. Again, we see a fairly low p-value (0.038) and a fairly high power (0.81), although there is more overlap between the two distributions (Fig. \@ref(fig:pmcResults)D).

```{r}
## The true value of the test statistic
delta <- -2*(results[["mpd.OUM"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)

## The distribution of the test statistic when the data is generated by the oum.mpd model
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oum.mpd.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oum.mpd.data"]]$loglik) %>% unlist))

## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpd.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```

Finally, we can look at the p-value and the power for the fifth comparison, between the OUMV and OUM models fit to the mpdMpleth hypothesis. Again, we see a fairly low p-value (0.046) and a fairly high power (0.8), although there is clearly more overlap between the two distributions  (Fig. \@ref(fig:pmcResults)E).

```{r}
## The true value of the test statistic
delta <- -2*(results[["mpdMpleth.OUM"]]$loglik - results[["mpdMpleth.OUMV"]]$loglik)

## The distribution of the test statistic when the data is generated by the oum.mpdMpleth model
delta.dist <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpdMpleth.to.oum.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oum.mpdMpleth.data"]]$loglik) %>% unlist))

## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits, function(f) f[["fit.oum.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist) - 
                    (lapply(PMCfits, function(f) f[["fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data"]]$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```

So, based on the above, I am creating a new regime painting that separates out the two metamorphosers from the non-metamorphosers, but keeps paedomorphs and direct developers in the same regime. I'm going to call this regime painting meta2.

```{r, echo=FALSE, fig.height=3, fig.width=5, fig.cap="A new evolutionary hypothesis that separates out feeding and non-feeding metamorphosers, but keeps all non-metamorphosers as a single strategy. Red = Metamorphosis (feeding), Green = Plethodontid metamorphosis (non-feeding), Cyan = Direct development or paedomorphosis."}
tree_meta2 <- tree_metamorphosis
data.meta2 <- data.metamorphosis
colnames(data.meta2)[2] <- "meta2"
## Separate out feeding and non-feeding metamorphosers using the mpdMpleth painting
tree_meta2$node.label[which(tree_meta2$node.label==2)] <- tree_mpdMpleth$node.label[which(tree_meta2$node.label==2)]
data.meta2[which(data.meta2$meta2==2),"meta2"] <- data.mpdMpleth[which(data.meta2$meta2==2),"mpdMpleth"]

par(mar=c(0.5,0.5,2,0.5), oma=c(0.5,0.5,0.5,0.5))
plot(tree_meta2)
nodelabels(pch=21, bg=tree_meta2$node.label)

```

```{r, echo=FALSE, eval=FALSE}
meta2.OUM <- OUwie(tree_meta2,data.meta2,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
meta2.BMS <- OUwie(tree_meta2,data.meta2,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
meta2.OUMA <- OUwie(tree_meta2,data.meta2,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
meta2.OUMV <- OUwie(tree_meta2,data.meta2,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
meta2.OUMVA <- OUwie(tree_meta2,data.meta2,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

results <- list(meta2.OUM=meta2.OUM,
                meta2.BMS=meta2.BMS,
                meta2.OUMA=meta2.OUMA,
                meta2.OUMV=meta2.OUMV,
                meta2.OUMVA=meta2.OUMVA
                )
saveRDS(results, "all_model_fits_novel_hyp.RDS")

```

If we fit this evolutionary hypothesis to the genome size data and compare the results with the pre-existing hypotheses, we see that it actually provides a better fit to the data than almost any other hypothesis, although the most complex hypothesis is still a better fit.

```{r}
results <- readRDS("all_model_fits.RDS")
results2 <- readRDS("all_model_fits_novel_hyp.RDS")
data.frame(model=names(c(sapply(results, function(x) x$AICc), sapply(results2, function(x) x$AICc))),
           "log L"=c(sapply(results, function(x) x$loglik), sapply(results2, function(x) x$loglik)),
           AICc=c(sapply(results, function(x) x$AICc), sapply(results2, function(x) x$AICc))) %>%
  arrange(., AICc)
  

```

```{r, echo=FALSE, eval=FALSE}
PMCdata2 <- list(length=500)
for (i in 1:500) {
  ## OUMV under the mpdMpleth hypothesis
  OUwie.sim(tree_mpdMpleth, 
            data.mpdMpleth, 
            shift.point=1, 
            alpha=results[["mpdMpleth.OUMV"]]$solution['alpha',] %>% as.numeric,
            sigma.sq=results[["mpdMpleth.OUMV"]]$solution['sigma.sq',] %>% as.numeric,
            theta0=results[["mpdMpleth.OUMV"]]$theta[2,1] %>% as.numeric,
            theta=results[["mpdMpleth.OUMV"]]$theta[,1] %>% as.numeric) -> oumv.mpdMpleth.data
  ## Note: OUwie.sim rewrites the names of the regimes, so fix that problem now or when you do any 
  ## refitting there will be problems (in particular, tree_mpdMpleth and oum.mpdMpleth.data have different
  ## names for the regimes represented on the internal nodes and the tips, respectively)
  oumv.mpdMpleth.data[,2] <- data.mpdMpleth[,2]
  ## OUMV under the meta2 hypothesis
  OUwie.sim(tree_meta2, 
            data.meta2, 
            shift.point=1, 
            alpha=results2[["meta2.OUMV"]]$solution['alpha',] %>% as.numeric,
            sigma.sq=results2[["meta2.OUMV"]]$solution['sigma.sq',] %>% as.numeric,
            theta0=results2[["meta2.OUMV"]]$theta[1,1] %>% as.numeric,
            theta=results2[["meta2.OUMV"]]$theta[,1] %>% as.numeric) -> oumv.meta2.data
  oumv.meta2.data[,2] <- data.meta2[,2]
  PMCdata2[[i]] <- list(oumv.mpdMpleth.data=oumv.mpdMpleth.data,
                        oumv.meta2.data=oumv.meta2.data)
}

fitComparison2 <- function(data) {
  
  ## Fit the meta2 model to the oumv.mpdMpleth data
  mpdMpleth.as.meta2 <- data[["oumv.mpdMpleth.data"]]
  mpdMpleth.as.meta2[,2] <- data[["oumv.meta2.data"]][,2]
  fit.oumv.meta2.to.oumv.mpdMpleth.data <- OUwie(tree_meta2, mpdMpleth.as.meta2, model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpdMpleth model to the oumv.meta2 data
  meta2.as.mpdMpleth <- data[["oumv.meta2.data"]]
  meta2.as.mpdMpleth[,2] <- data[["oumv.mpdMpleth.data"]][,2]
  fit.oumv.mpdMpleth.to.oumv.meta2.data <- OUwie(tree_mpdMpleth, meta2.as.mpdMpleth, model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the mpdMpleth model to the oumv.mpdMpleth data
  fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data <- OUwie(tree_mpdMpleth,data[["oumv.mpdMpleth.data"]],model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  ## Fit the meta2 model to the oumv.meta2 data
  fit.oumv.meta2.to.oumv.meta2.data <- OUwie(tree_meta2,data[["oumv.meta2.data"]],model=c("OUMV"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
  
  #########################################################################

  list(fit.oumv.meta2.to.oumv.mpdMpleth.data=fit.oumv.meta2.to.oumv.mpdMpleth.data,
       fit.oumv.mpdMpleth.to.oumv.meta2.data=fit.oumv.mpdMpleth.to.oumv.meta2.data,
       fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data=fit.oumv.mpdMpleth.to.oumv.mpdMpleth.data,
       fit.oumv.meta2.to.oumv.meta2.data=fit.oumv.meta2.to.oumv.meta2.data
       )
}

mclapply(PMCdata2,
         fitComparison2,
         mc.cores=4) -> PMCfits2

```

#### Analyses with a different evolutionary history for plethodontids
In the above analyses, the ancestor of all plethodontids is assumed to be a metamorphosing amphibian. However, there is some debate about this, and some researchers think that this ancestor was actually a direct developer. As such, we want to make sure that the conlusions drawn above are unaffected by this assumption. This requires modifying all three adaptive hypotheses to "repaint" the ancestral node.

```{r, echo=FALSE}
# which(tree_mpdMpleth$node.label=="3")
# [1] 23 40 41 42 45 46 81 82 83 85 86 87 88
tree_mpdMpleth2 <- tree_mpdMpleth
tree_mpdMpleth2$node.label[23] <- "1"

tree_mpd2 <- tree_mpd
tree_mpd2$node.label[23] <- "1"

tree_metamorphosis2 <- tree_metamorphosis
tree_metamorphosis2$node.label[23] <- "5"

# plot(tree_mpdMpleth2)
# nodelabels(pch=21, bg=tree_mpdMpleth2$node.label)
# 
# plot(tree_mpd2)
# nodelabels(pch=21, bg=tree_mpd2$node.label)
# 
# plot(tree_metamorphosis2)
# nodelabels(pch=21, bg=tree_metamorphosis2$node.label)

```


```{r, echo=FALSE, eval=FALSE}
## Fit all of the model variants (BMS, OUMA, OUMV, OUMVA) to each hypothesis
metamorphosis.OUM <- OUwie(tree_metamorphosis2,data.metamorphosis,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.BMS <- OUwie(tree_metamorphosis2,data.metamorphosis,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMA <- OUwie(tree_metamorphosis2,data.metamorphosis,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMV <- OUwie(tree_metamorphosis2,data.metamorphosis,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMVA <- OUwie(tree_metamorphosis2,data.metamorphosis,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

mpd.OUM <- OUwie(tree_mpd2,data.mpd,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.BMS <- OUwie(tree_mpd2,data.mpd,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMA <- OUwie(tree_mpd2,data.mpd,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMV <- OUwie(tree_mpd2,data.mpd,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMVA <- OUwie(tree_mpd2,data.mpd,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

mpdMpleth.OUM <- OUwie(tree_mpdMpleth2,data.mpdMpleth,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.BMS <- OUwie(tree_mpdMpleth2,data.mpdMpleth,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMA <- OUwie(tree_mpdMpleth2,data.mpdMpleth,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMV <- OUwie(tree_mpdMpleth2,data.mpdMpleth,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMVA <- OUwie(tree_mpdMpleth2,data.mpdMpleth,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

## save these results
results <- list(metamorphosis.OUM=metamorphosis.OUM,
                metamorphosis.BMS=metamorphosis.BMS,
                metamorphosis.OUMA=metamorphosis.OUMA,
                metamorphosis.OUMV=metamorphosis.OUMV,
                metamorphosis.OUMVA=metamorphosis.OUMVA,
                mpd.OUM=mpd.OUM,
                mpd.BMS=mpd.BMS,
                mpd.OUMA=mpd.OUMA,
                mpd.OUMV=mpd.OUMV,
                mpd.OUMVA=mpd.OUMVA,
                mpdMpleth.OUM=mpdMpleth.OUM,
                mpdMpleth.BMS=mpdMpleth.BMS,
                mpdMpleth.OUMA=mpdMpleth.OUMA,
                mpdMpleth.OUMV=mpdMpleth.OUMV,
                mpdMpleth.OUMVA=mpdMpleth.OUMVA
                )
saveRDS(results, "all_model_fits_2.RDS")

```

```{r, echo=FALSE}
results <- readRDS("all_model_fits_2.RDS")
```

You can see that the order of the models is completely unaffected by this change, which makes sense, given how ancient this branching point is. The best-fitting model remains the OU model with 4 selective regimes (1=direct development, 2=metamorphosis, 3=metamorphosing plethodontids, 4=paedomorphosis), with separate drift parameters for each regime, although the model with a single drift parameter has very similar AICc score. Here are the AICc values and parameter estimates for each of the models.

```{r, echo=FALSE}
data.frame(Hypothesis = lapply(results, function(x) x$AICc) %>% unlist %>% names, 
           AICc = lapply(results, function(x) x$AICc) %>% unlist %>% signif(3),
           alpha = lapply(results, function(x) unique(x$solution['alpha',]) %>% signif(3)) %>% unlist,
           sigma.sq = lapply(results, function(x) unique(x$solution['sigma.sq',]) %>% signif(3) %>% paste(., collapse=", ")) %>% unlist,
           theta = lapply(results, function(x) x$theta[,1] %>% signif(3) %>% paste(., collapse=", ")) %>% unlist
           ) %>% 
  arrange(AICc) %>% as.data.frame
```


#### Analyses with $\texttt{ouch}$
One final caveat: in all of the above, I fit the models using the $\texttt{OUwie}$ option 'root.station=FALSE', which means that the value at the base of the tree is not drawn from the stationary distribution of that regime. However, the root state is not estimated as a separate optimum either - instead the root is treated as any other node in its regime. This is because it was unclear how to modify the covariance matrix to account for the root being drawn from the stationary distribution when there are multiple $\alpha$ or $\sigma$ values. Neither Jeremy Beaulieu nor Lam Si Tung Ho knew how to modify it for these complex models. Therefore, to make the comparison fair, I set root.station=FALSE for *all* of the models (even though I could have set root.station=TRUE for the OUM models). Unfortunately, this is not a trivial change, as you can see by comparing the parameter estimates and likelihoods under the two models. In particular, as soon as root.station is set to false, the estimates of selection strength and the optima jump to fairly ridiculous values, but the likelihood also improves considerably.

```{r, eval=FALSE}
## The fit when root.station=TRUE
mpdMpleth.ouwie.1 <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.ouwie.1

## The fit when root.station=FALSE
mpdMpleth.ouwie.2 <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.ouwie.2

```

Therefore, I feel that it is useful to actually repeat all of the analyses with $\texttt{ouch}$, just to make sure that the overall phylogenetic conclusions don't change, even though the parameter estimates do. One major difference is that Brownian motion fits the data much better now than it did under the $\texttt{OUwie}$ simulations.

```{r}
mpdMpleth.ouch <- hansen(gsz, tree, regimes['mpdMpleth'], sqrt.alpha=1, sigma=1)
mpd.ouch <- hansen(gsz, tree, regimes['mpd'], sqrt.alpha=1, sigma=1)
meta.ouch <- hansen(gsz, tree, regimes['metamorphosis'], sqrt.alpha=1, sigma=1)
ou1.ouch <- hansen(gsz, tree, regimes['OU1'], sqrt.alpha=1, sigma=1)
bm.ouch <- brown(gsz, tree)

data.frame(Hypothesis=c("mpdMpleth","mpd","metamorphosis","ou1","bm"),
           AICc=c(summary(mpdMpleth.ouch)$aic.c %>% signif(3),
                  summary(mpd.ouch)$aic.c %>% signif(3),
                  summary(meta.ouch)$aic.c %>% signif(3),
                  summary(ou1.ouch)$aic.c %>% signif(3),
                  summary(bm.ouch)$aic.c) %>% signif(3),
           alpha=c(summary(mpdMpleth.ouch)$alpha %>% as.numeric() %>% signif(3),
                  summary(mpd.ouch)$alpha %>% as.numeric() %>% signif(3),
                  summary(meta.ouch)$alpha %>% as.numeric() %>% signif(3),
                  summary(ou1.ouch)$alpha %>% as.numeric() %>% signif(3),
                  NA),
          sigma.sq=c(summary(mpdMpleth.ouch)$sigma.sq %>% as.numeric() %>% signif(3),
                  summary(mpd.ouch)$sigma.sq %>% as.numeric() %>% signif(3),
                  summary(meta.ouch)$sigma.sq %>% as.numeric() %>% signif(3),
                  summary(ou1.ouch)$sigma.sq %>% as.numeric() %>% signif(3),
                  summary(bm.ouch)$sigma.sq %>% as.numeric() %>% signif(3)),
          theta=c(summary(mpdMpleth.ouch)$optima$genomesize %>% signif(3) %>% paste(., collapse=", "),
                  summary(mpd.ouch)$optima$genomesize %>% signif(3) %>% paste(., collapse=", "),
                  summary(meta.ouch)$optima$genomesize %>% signif(3) %>% paste(., collapse=", "),
                  summary(ou1.ouch)$optima$genomesize %>% signif(3) %>% paste(., collapse=", "),
                  NA)) %>% arrange(., AICc)
```
We can also compare the evolutionary conclusions that we would draw from the parameter values themselves to those they we drew above with the OUMV model. You can see from the table below that the analysis results in very similar conclusions: for direct developers, there is a weak deterministic trend towards smaller genomes, but the stochastic component of the trait dynamics is larger; for metamorphosers, trait change is dominated by random processes because the mean genome size is close to the optimum; for metamorphosing plethodontids, there is a strong deterministic trend for small genomes; for paedomorphs, there is a strong deterministic trend towards large genomes. 

```{r}
data.mpdMpleth %>% group_by(mpdMpleth) %>% summarise(meanGenomeSize=signif(mean(genomesize),3)) %>% 
  mutate(., 
         alpha=summary(mpdMpleth.ouch)$alpha %>% as.numeric() %>% signif(3),
         theta=summary(mpdMpleth.ouch)$optima$genomesize %>% signif(3),
         'selection strength'=(alpha*(theta-meanGenomeSize)) %>% signif(3),
         'drift intensity'=sqrt(summary(mpdMpleth.ouch)$sigma.sq %>% as.numeric() %>% signif(3))
  )

```
However, there is an inherent flaw in the above comparison, in that $\alpha$ and $\sigma^2$ have very different units, so that $\alpha (\theta-X)$ has units of trait times time to the -1 power, whereas $\sigma$ has units of trait times time to the -1/2 power. 
As such, we can also compute the dimensionless parameters of Cressler et al. (2015), which were shown to be very good determinants of the power and parameter estimation capacities of the OU method.
I'm unsure of how to do a similar analysis with the OUMV model, since there are multiple $\sigma$ parameters.
Regardless: the two parameters identified in Cressler et al. (2015) were the discriminability ratio $\frac{\sqrt{2\alpha}\Delta\theta}{\sigma}$ and the selection opportunity $\alpha T$, where $\Delta\theta=3.62-3.41=0.21$ is the smallest difference between selective optima and $T=1$ is the depth of the tree.
Here the value of the discriminability ratio is `r sqrt(2*0.0941)*0.21/0.3478505` and the selection opportunity is just is 0.0941.
Both of these values are quite small, and Cressler et al. (2015) suggests that both power and parameter estimation capability are fairly low in this region of parameter space.
Thus, the results of the phylogenetic Monte Carlo are quite important, but very little should probably be made of the parameter estimates themselves.

Repeating all of the phylogenetic Monte Carlo to ensure that the conclusions still hold for the three comparisons above: adaptive versus non-adaptive, mpdMpleth versus mpd, and mpdMpleth versus metamorphosis.   

```{r, eval=FALSE, echo=FALSE}

mpdMpleth.simdata <- simulate(mpdMpleth.ouch, nsim=500)
mpd.simdata <- simulate(mpd.ouch, nsim=500)
meta.simdata <- simulate(meta.ouch, nsim=500)
bm.simdata <- simulate(bm.ouch, nsim=500)

## Comparison 1
mpdMpleth.fit.to.mpdMpleth.data <- lapply(mpdMpleth.simdata, function(x) hansen(x, tree, regimes['mpdMpleth'], sqrt.alpha=0.1, sigma=0.1))
bm.fit.to.mpdMpleth.data <- lapply(mpdMpleth.simdata, function(x) brown(x, tree))
bm.fit.to.bm.data <- lapply(bm.simdata, function(x) brown(x, tree))
mpdMpleth.fit.to.bm.data <- lapply(bm.simdata, function(x) hansen(x, tree, regimes['mpdMpleth'], sqrt.alpha=0.1, sigma=0.1))

## Comparison 2
mpd.fit.to.mpdMpleth.data <- lapply(mpdMpleth.simdata, function(x) hansen(x, tree, regimes['mpd'], sqrt.alpha=0.1, sigma=0.1))
mpd.fit.to.mpd.data <- lapply(mpd.simdata, function(x) hansen(x, tree, regimes['mpd'], sqrt.alpha=0.1, sigma=0.1))
mpdMpleth.fit.to.mpd.data <- lapply(mpd.simdata, function(x) hansen(x, tree, regimes['mpdMpleth'], sqrt.alpha=0.1, sigma=0.1))

## Comparison 3
meta.fit.to.mpdMpleth.data <- lapply(mpdMpleth.simdata, function(x) hansen(x, tree, regimes['metamorphosis'], sqrt.alpha=0.1, sigma=0.1))
meta.fit.to.meta.data <- lapply(meta.simdata, function(x) hansen(x, tree, regimes['metamorphosis'], sqrt.alpha=0.1, sigma=0.1))
mpdMpleth.fit.to.meta.data <- lapply(meta.simdata, function(x) hansen(x, tree, regimes['mpdMpleth'], sqrt.alpha=0.1, sigma=0.1))

PMCfits2 <- list(mpdMpleth.fit.to.mpdMpleth.data=mpdMpleth.fit.to.mpdMpleth.data,
                bm.fit.to.mpdMpleth.data=bm.fit.to.mpdMpleth.data,
                bm.fit.to.bm.data=bm.fit.to.bm.data,
                mpdMpleth.fit.to.bm.data=mpdMpleth.fit.to.bm.data,
                mpd.fit.to.mpdMpleth.data=mpd.fit.to.mpdMpleth.data,
                mpd.fit.to.mpd.data=mpd.fit.to.mpd.data,
                mpdMpleth.fit.to.mpd.data=mpdMpleth.fit.to.mpd.data,
                meta.fit.to.mpdMpleth.data=meta.fit.to.mpdMpleth.data,
                meta.fit.to.meta.data=meta.fit.to.meta.data,
                mpdMpleth.fit.to.meta.data=mpdMpleth.fit.to.meta.data
                )
saveRDS(PMCfits2, "PMCfits2.RDS")

```


For the comparison of the mpdMpleth hypothesis to the non-adaptive hypothesis, the p-value is 0.046 and the power is 0.82.
```{r}
PMCfits2 <- readRDS("PMCfits2.RDS")
## The true value of the test statistic
delta <- -2*(summary(bm.ouch)$loglik - summary(mpdMpleth.ouch)$loglik)

## The distribution of the test statistic when the data is generated by the oum.mpd model
delta.dist <- -2*((lapply(PMCfits2[["bm.fit.to.bm.data"]], function(f) summary(f)$loglik) %>% unlist) - 
                    (lapply(PMCfits2[["mpdMpleth.fit.to.bm.data"]], function(f) summary(f)$loglik) %>% unlist))

## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits2[["bm.fit.to.mpdMpleth.data"]], function(f) summary(f)$loglik) %>% unlist) - 
                    (lapply(PMCfits2[["mpdMpleth.fit.to.mpdMpleth.data"]], function(f) summary(f)$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```
```{r bmVSmpdMpleth, echo=FALSE, fig.cap="Comparing the distributions of the test statistic when the data is generated by a BM model (BM) or an OU model under the mpdMpleth hypothesis (mpdMpleth). The dashed line gives the actual value of the test statistic for the true data."}
par(mar=c(5,5,1,1), oma=rep(0,4))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('BM','mpdMpleth'), fill=c(gray(0.7),gray(0.2)), bty='n')

```

For the comparison of the mpdMpleth hypothesis to the mpd hypothesis, the p-value is 0.052 and the power is 0.64, so we are a bit less confident that we can really discriminate between these two hypotheses.
```{r}
## The true value of the test statistic
delta <- -2*(summary(mpd.ouch)$loglik - summary(mpdMpleth.ouch)$loglik)

## The distribution of the test statistic when the data is generated by the oum.mpd model
delta.dist <- -2*((lapply(PMCfits2[["mpd.fit.to.mpd.data"]], function(f) summary(f)$loglik) %>% unlist) - 
                    (lapply(PMCfits2[["mpdMpleth.fit.to.mpd.data"]], function(f) summary(f)$loglik) %>% unlist))

## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits2[["mpd.fit.to.mpdMpleth.data"]], function(f) summary(f)$loglik) %>% unlist) - 
                    (lapply(PMCfits2[["mpdMpleth.fit.to.mpdMpleth.data"]], function(f) summary(f)$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```
```{r mpdVSmpdMpleth, echo=FALSE, fig.cap="Comparing the distributions of the test statistic when the data is generated by an OU model under the mpd hypothesis (mpd) or an OU model under the mpdMpleth hypothesis (mpdMpleth). The dashed line gives the actual value of the test statistic for the true data."}
par(mar=c(5,5,1,1), oma=rep(0,4))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('mpd','mpdMpleth'), fill=c(gray(0.7),gray(0.2)), bty='n')

```

For the comparison of the mpdMpleth hypothesis to the metamorphosis hypothesis, the p-value is 0.072 and the power is 0.73, so we are a bit less confident that we can really discriminate between these two hypotheses.
```{r}
## The true value of the test statistic
delta <- -2*(summary(meta.ouch)$loglik - summary(mpdMpleth.ouch)$loglik)

## The distribution of the test statistic when the data is generated by the oum.mpd model
delta.dist <- -2*((lapply(PMCfits2[["meta.fit.to.meta.data"]], function(f) summary(f)$loglik) %>% unlist) - 
                    (lapply(PMCfits2[["mpdMpleth.fit.to.meta.data"]], function(f) summary(f)$loglik) %>% unlist))

## p-value for the test
sum(delta < delta.dist)/length(delta.dist)

## The distribution of the test statistic when the data is generated by the oumv.mpdMpleth model
delta.dist.2 <- -2*((lapply(PMCfits2[["meta.fit.to.mpdMpleth.data"]], function(f) summary(f)$loglik) %>% unlist) - 
                    (lapply(PMCfits2[["mpdMpleth.fit.to.mpdMpleth.data"]], function(f) summary(f)$loglik) %>% unlist))
## Power of the test
sum(delta.dist.2 > sort(delta.dist)[475])/length(delta.dist.2)

```
```{r metaVSmpdMpleth, echo=FALSE, fig.cap="Comparing the distributions of the test statistic when the data is generated by an OU model under the metamorphosis hypothesis (metamorph) or an OU model under the mpdMpleth hypothesis (mpdMpleth). The dashed line gives the actual value of the test statistic for the true data."}
par(mar=c(5,5,1,1), oma=rep(0,4))
d1 <- density(delta.dist)
d2 <- density(delta.dist.2)
plot.new()
plot.window(xlim=range(c(d1$x,d2$x)), ylim=range(c(d1$y,d2$y)))
axis(1);axis(2);box('plot')
mtext(side=1, line=3, expression(delta))
mtext(side=2, line=3, 'Density')
lines(d1$x, d1$y, col=gray(0.7))
with(d1, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.7, alpha=0.5)))
lines(d2$x, d2$y, col=gray(0.2))
with(d2, polygon(x=c(min(x), x[x<max(x)], rev(x[x<max(x)]), min(x)), y=c(0, 0*x[x<max(x)], rev(y[x<max(x)]), 0), col=gray(0.2, alpha=0.5)))
abline(v=delta, lwd=2, lty=2)
legend(x='topright', c('metamorph','mpdMpleth'), fill=c(gray(0.7),gray(0.2)), bty='n')

```
