---
title: "Phylogenetic Analysis of Salamander Genome Size"
author: "Clay Cressler"
date: "02/06/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

This is an updated version of the analyses that were initially carried out several years ago. 
We had wanted to use $\texttt{Ouwie}$ to do these analyses but previously, there were undiagnosed issues with its fitting algorithm.
As such, we ended up using $\texttt{ouch}$ for all analyses, which prevented us from ever testing models with multiple $\alpha$ or $\sigma^2$ parameters.
However, I spent time working through these issues in Spring 2020 and am now satisfied that $\texttt{ouch}$ and $\texttt{OUwie}$ produce identical results, at least once you specify a value for 'shift.point' to make the way that $\texttt{OUwie}$ handles regimes identical to $\texttt{ouch}$.
Nevertheless, I will fit the models using both $\texttt{OUwie}$ and $\texttt{ouch}$ to confirm they are giving similar results when they should. 

```{r, echo=FALSE}
library(ape, quietly=TRUE)
library(ouch, quietly=TRUE)
library(OUwie, quietly=TRUE)
library(tidyverse)

tree <- read.nexus("av_ultra_fulldataset.nex")
tree_metamorphosis <- tree_mpd <- tree_mpdMpleth <- tree
## Load the hypotheses
nodedata <- read.csv('ouwie_nodelabels.csv')
nodedata <- nodedata[,c('node','metamorphosis','mpd','mpdMpleth')]
## metamorphosis has two regimes: M, x
## mpd has three regimes: D, M, P
## mpdMpleth has four regimes: D, M, Mpleth, P
## Node Colors for plots
nodecol <- nodedata
nodecol[-1] <- sapply(nodecol[-1], as.character)
nodecol[nodecol=="x"] <- "brown"
nodecol[nodecol=="M"] <- "red"
nodecol[nodecol=="D"] <- "black"
nodecol[nodecol=="P"] <- "dodgerblue1"
nodecol[nodecol=="Mpleth"] <- "purple"
## code Regime names as numbers for the OUwie algorithm
nn <- nodedata
nn[-1] <- sapply(nn[-1], as.character)
nn[nn=="D"] <- 1
nn[nn=="M"] <- 2
nn[nn=="Mpleth"] <- 3
nn[nn=="P"] <- 4
nn[nn=="x"] <- 5
## Add node labels to trees
tree_metamorphosis$node.label <- nn$metamorphosis
tree_mpd$node.label <- nn$mpd
tree_mpdMpleth$node.label <- nn$mpdMpleth
## Tip data
dat <- read.csv("tree1.dat.csv")
dat$genomesize <- log(dat$genomesize)
tdat <- dat[!is.na(dat$genomesize),]  # tip data
ndat <- dat[is.na(dat$genomesize),]   # node data
## reorder tdat to match order of tip labels in phylo tree
oo <- sapply(tree$tip.label, function(x) grep(x, tdat$labels))
tipdat <- tdat[oo,]
tipdat <- tipdat[c("labels", "metamorphosis", "mpd", "mpdMpleth", "genomesize")]
## Tip Colors for plots
tipcol <- tipdat
tipcol[,1:(ncol(tipdat)-1)] <- sapply(tipcol[,1:(ncol(tipdat)-1)], as.character)
tipcol[tipcol=="x"] <- "brown"
tipcol[tipcol=="M"] <- "red"
tipcol[tipcol=="D"] <- "black"
tipcol[tipcol=="P"] <- "dodgerblue1"
tipcol[tipcol=="Mpleth"] <- "purple"
## code Regime names as numbers for the OUwie algorithm
tt <- tipdat
tt[,1:(ncol(tipdat)-1)] <- sapply(tt[,1:(ncol(tipdat)-1)], as.character)
tt[tt=="D"] <- 1
tt[tt=="M"] <- 2
tt[tt=="Mpleth"] <- 3
tt[tt=="P"] <- 4
tt[tt=="x"] <- 5

### input dataframes - one for each hypothesis
data.metamorphosis <- tt[c("labels", "metamorphosis", "genomesize")]
data.mpd <- tt[c("labels", "mpd", "genomesize")]
data.mpdMpleth <- tt[c("labels", "mpdMpleth", "genomesize")]


##############################################################
## OUCH setup
##############################################################
dat1 <- read.csv("tree1.dat.csv")
tree1 <- read.nexus("av_ultra_fulldataset.nex")
tree <- ape2ouch(tree1)
dat <- dat1
rownames(dat) <- dat$nodes
regimes <- dat[c("metamorphosis", "mpd", "mpdMpleth","OU1","fivereg")]
gsz <- log(dat['genomesize'])
```

First up is a simple comparison of Brownian motion fits, which you can see are almost identical. 

```{r, echo=TRUE}
## Single-rate BM (this is hypothesis-independent)
bm.ouwie <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("BM1"),root.station=TRUE,scaleHeight=TRUE,quiet=TRUE)
bm.ouch <- brown(gsz, tree)

summary(bm.ouch)$sigma.squared %>% as.numeric
bm.ouwie$solution["sigma.sq",]

summary(bm.ouch)$theta %>% as.numeric
bm.ouwie$theta[1]

```
```{r fig:mpdMpleth, fig.height=6, fig.width=6, fig.cap="The mpdMpleth regime painting"}
plot(tree_mpdMpleth)
nodelabels(pch=21, bg=tree_mpdMpleth$node.label)

```
Now let's fit a multiple-optima Ornstein-Uhlenbeck model to the mpdMpleth regime painting using both $\texttt{OUwie}$ and $\texttt{ouch}$, shown in Fig. \@ref(fig:mpdMpleth).
From this, you can see that all of the parameter estimates are quite similar, and the likelihood, calculated by $\texttt{ouch}$ using the $\texttt{OUwie}$ parameter estimates, is identical to the fourth decimal place. 
Obviously I wish they were completely identical, but I'll take it.

```{r, echo=TRUE}
## Fit the mpdMpleth hypothesis
mpdMpleth.ouwie <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.ouch <- hansen(gsz, tree, regimes['mpdMpleth'], sqrt.alpha=1, sigma=1)

## Compare sigma.sq estimates
summary(mpdMpleth.ouch)$sigma.squared %>% as.numeric
mpdMpleth.ouwie$solution["sigma.sq",1]

## Compare alpha estimates
summary(mpdMpleth.ouch)$alpha %>% as.numeric
mpdMpleth.ouwie$solution["alpha",1]

## Compare optima estimates
summary(mpdMpleth.ouch)$optima$genomesize
mpdMpleth.ouwie$theta[,1]

## Compare the likelihoods
summary(hansen(gsz, tree=tree, regimes=regimes['mpdMpleth'], sqrt.alpha=sqrt(mpdMpleth.ouwie$solution["alpha",1]), sigma=sqrt(mpdMpleth.ouwie$solution["sigma.sq",1]), fit=FALSE))$loglik
mpdMpleth.ouwie$loglik


```

I will repeat this analysis with both methods for the other regime paintings, starting with the mpd painting shown in Fig. \@ref(fig:mpd).
And again, we see really strong evidence that the two methods are nearly identical.

```{r fig:mpd, fig.height=6, fig.width=6, fig.cap="The mpd regime painting"}
plot(tree_mpd)
nodelabels(pch=21, bg=tree_mpd$node.label)

```


```{r, echo=TRUE}
## Fit the mpd hypothesis
mpd.ouwie <- OUwie(tree_mpd,data.mpd,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.ouch <- hansen(gsz, tree, regimes['mpd'], sqrt.alpha=1, sigma=1)

## Compare sigma.sq estimates
summary(mpd.ouch)$sigma.squared %>% as.numeric
mpd.ouwie$solution["sigma.sq",1]

## Compare alpha estimates
summary(mpd.ouch)$alpha %>% as.numeric
mpd.ouwie$solution["alpha",1]

## Compare optima estimates
summary(mpd.ouch)$optima$genomesize
mpd.ouwie$theta[,1]

## Compare the likelihoods
summary(hansen(gsz, tree=tree, regimes=regimes['mpd'], sqrt.alpha=sqrt(mpd.ouwie$solution["alpha",1]), sigma=sqrt(mpd.ouwie$solution["sigma.sq",1]), fit=FALSE))$loglik
mpd.ouwie$loglik
```

Finally, I will consider the 'metamorphosis' tree painting shown in Fig. \@ref(fig:meta), where I again see good agreement between $\texttt{OUwie}$ and $\texttt{ouch}$.

```{r fig:meta, fig.height=6, fig.width=6, fig.cap="The metamorphosis regime painting"}
plot(tree_metamorphosis)
nodelabels(pch=21, bg=tree_metamorphosis$node.label)

```

```{r, echo=TRUE}
## Fit the metamorphosis hypothesis
metamorphosis.ouwie <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUM"),root.station=TRUE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.ouch <- hansen(gsz, tree, regimes['metamorphosis'], sqrt.alpha=1, sigma=1)

## Compare sigma.sq estimates
summary(metamorphosis.ouch)$sigma.squared %>% as.numeric
metamorphosis.ouwie$solution["sigma.sq",1]

## Compare alpha estimates
summary(metamorphosis.ouch)$alpha %>% as.numeric
metamorphosis.ouwie$solution["alpha",1]

## Compare optima estimates
summary(metamorphosis.ouch)$optima$genomesize
metamorphosis.ouwie$theta[,1]

## Compare the likelihoods
summary(hansen(gsz, tree=tree, regimes=regimes['metamorphosis'], sqrt.alpha=sqrt(metamorphosis.ouwie$solution["alpha",1]), sigma=sqrt(metamorphosis.ouwie$solution["sigma.sq",1]), fit=FALSE))$loglik
metamorphosis.ouwie$loglik
```
As such, I will proceed with fitting all of the other multi-rate models: a Brownian motion model where the drift parameter ($\sigma$) varies across regimes (BMS), an Ornstein-Uhlenbeck (OU) model where selection strength varies across regimes ($\alpha$) (OUMA), an OU model where drift varies across regimes (OUMV), and an OU model where both selection and drift vary across regimes (OUMVA). 

```{r, echo=TRUE, eval=FALSE}
## Fit all of the model variants (BMS, OUMA, OUMV, OUMVA) to each hypothesis
metamorphosis.OUM <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.BMS <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMA <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMV <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
metamorphosis.OUMVA <- OUwie(tree_metamorphosis,data.metamorphosis,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

mpd.OUM <- OUwie(tree_mpd,data.mpd,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.BMS <- OUwie(tree_mpd,data.mpd,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMA <- OUwie(tree_mpd,data.mpd,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMV <- OUwie(tree_mpd,data.mpd,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpd.OUMVA <- OUwie(tree_mpd,data.mpd,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

mpdMpleth.OUM <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUM"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.BMS <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("BMS"),root.station=FALSE, scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMA <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUMA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMV <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUMV"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)
mpdMpleth.OUMVA <- OUwie(tree_mpdMpleth,data.mpdMpleth,model=c("OUMVA"), scaleHeight=TRUE, shift.point=1, quiet=TRUE)

## save these results
results <- list(BM1=bm.ouwie,
                metamorphosis.OUM=metamorphosis.OUM,
                metamorphosis.BMS=metamorphosis.BMS,
                metamorphosis.OUMA=metamorphosis.OUMA,
                metamorphosis.OUMV=metamorphosis.OUMV,
                metamorphosis.OUMVA=metamorphosis.OUMVA,
                mpd.OUM=mpd.OUM,
                mpd.BMS=mpd.BMS,
                mpd.OUMA=mpd.OUMA,
                mpd.OUMV=mpd.OUMV,
                mpd.OUMVA=mpd.OUMVA,
                mpdMpleth.OUM=mpdMpleth.OUM,
                mpdMpleth.BMS=mpdMpleth.BMS,
                mpdMpleth.OUMA=mpdMpleth.OUMA,
                mpdMpleth.OUMV=mpdMpleth.OUMV,
                mpdMpleth.OUMVA=mpdMpleth.OUMVA
                )
saveRDS(results, "all_model_fits.RDS")

```

```{r}
results <- readRDS("all_model_fits.RDS")
```

The best-fitting model is the OU model with 4 selective regimes (1=direct development, 2=metamorphosis, 3=metamorphosing plethodontids, 4=plethodontids), with separate drift parameters for each regime, although the model with a single drift parameter has very similar AICc score.

```{r, echo=TRUE}
data.frame(Hypothesis = lapply(results, function(x) x$AICc) %>% unlist %>% names, 
           AICc = lapply(results, function(x) x$AICc) %>% unlist) %>% 
  arrange(AICc)
```

To help interpret the biological implications of the parameter values of the best-fitting model, it is useful to calculate the values of the selection strength and drift terms of the OU model ($\alpha*(\theta-X)$ and $\sigma$).
I will assume that $X$ is the average genome size for salamanders in each of the selective regimes. 
Plugging those genome size values, along with the corresponding parameter estimates for each regime gives a sense of how strong selection and drift are in each regime.
Based on this data, you can say genome size for the direct developers and the metamorphosers is primarily driven by drift, whereas plethodontids appear to be under fairly strong selection for either small genome size (metamorphosing plethodontids) or large genome size (direct-developing plethodontids).

```{r, echo=TRUE}
data.mpdMpleth %>% group_by(mpdMpleth) %>% summarise(meanGenomeSize=signif(mean(genomesize),3)) %>% 
  mutate(., 
         alpha=results[["mpdMpleth.OUMV"]]$solution['alpha',] %>% signif(3),
         theta=results[["mpdMpleth.OUMV"]]$theta[,1] %>% signif(3),
         'selection strength'=(alpha*(theta-meanGenomeSize)) %>% signif(3),
         'drift intensity'=sqrt(results[["mpdMpleth.OUMV"]]$solution['sigma.sq',]) %>% signif(3))
                    
```

However, it is important to assess how strong the support for this hypothesis truly is. 
Following the protocol of Boettiger et al. (2012) for phylogenetic Monte Carlo, I need to simulate datasets under different models and ask how often the true (data-generating) model is rejected in favor of an alternative model.
At the very least, I need to do this for the multi-$\sigma$ and single-$\sigma$ OU models with four selective optima, but I should also include Brownian motion to include a non-adaptive hypothesis.
